File: packages/tactical-engine-py/tactical_engine/targeting/resolve_sphere.py
from shared_contracts import Coordinate, Obstacle
from ..grid.coordinates import chebyshev_distance
from ..validation.obstacle_rules import blocks_targeting


def resolve_sphere(
    center: Coordinate,
    radius: int,
    obstacles: list[Obstacle],
) -> list[Coordinate]:
    result: list[Coordinate] = []
    for x in range(center.x - radius, center.x + radius + 1):
        for y in range(center.y - radius, center.y + radius + 1):
            coordinate = Coordinate(x=x, y=y)
            if chebyshev_distance(center, coordinate) <= radius and not blocks_targeting(obstacles, coordinate):
                result.append(coordinate)
    return result

File: packages/tactical-engine/src/targeting/resolve-sphere.js
import { chebyshevDistance } from "../grid/coordinates";
import { blocksEffect } from "../validation/obstacle-rules";
import { canEffectReachAoE } from "./aoe-filter";
export function resolveSphere(center, radius, obstacles, edgeObstacles = []) {
    const result = [];
    for (let x = center.x - radius; x <= center.x + radius; x += 1) {
        for (let y = center.y - radius; y <= center.y + radius; y += 1) {
            const coordinate = { x, y };
            if (chebyshevDistance(center, coordinate) <= radius &&
                !blocksEffect(obstacles, coordinate) &&
                canEffectReachAoE(center, coordinate, obstacles, edgeObstacles)) {
                result.push(coordinate);
            }
        }
    }
    return result;
}
